ingester/ingester_lifecycle.go

```diff
@@ -0,0 +1,419 @@
+		}
+
+		// We exist in the ring, so assume the ring is right and copy out tokens & state out of there.
+		i.state = ingesterDesc.State
+		for _, token := range ringDesc.Tokens {
+			if token.Ingester == i.id {
+				i.tokens = append(i.tokens, token.Token)
+			}
+		}
+
+		log.Infof("Existing entry found in ring with state=%s, tokens=%v.", i.state, i.tokens)
+		return ringDesc, true, nil
+	})
+}
+
+// autoJoin selects random tokens & moves state to ACTIVE
+func (i *Ingester) autoJoin() error {
+	return i.consul.CAS(ring.ConsulKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		var ringDesc *ring.Desc
+		if in == nil {
+			ringDesc = ring.NewDesc()
+		} else {
+			ringDesc = in.(*ring.Desc)
+		}
+
+		// At this point, we should not have any tokens, and we should be in PENDING state.
+
+		var takenTokens, myTokens []uint32
+		for _, token := range ringDesc.Tokens {
+			takenTokens = append(takenTokens, token.Token)
+			if token.Ingester == i.id {
+				myTokens = append(myTokens, token.Token)
+			}
+		}
+
+		if len(myTokens) > 0 {
+			log.Errorf("%d tokens already exist for this ingester - wasn't expecting any!", len(myTokens))
+		}
+
+		newTokens := ring.GenerateTokens(i.cfg.NumTokens-len(myTokens), takenTokens)
+		i.state = ring.ACTIVE
+		ringDesc.AddIngester(i.id, i.addr, newTokens, i.state)
+
+		tokens := append(myTokens, newTokens...)
+		sort.Sort(sortableUint32(tokens))
+		i.tokens = tokens
+
+		return ringDesc, true, nil
+	})
+}
+
+// updateConsul updates our entries in consul, heartbeating and dealing with
+// consul restarts.
+func (i *Ingester) updateConsul() error {
+	return i.consul.CAS(ring.ConsulKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		var ringDesc *ring.Desc
+		if in == nil {
+			ringDesc = ring.NewDesc()
+		} else {
+			ringDesc = in.(*ring.Desc)
+		}
+
+		ingesterDesc, ok := ringDesc.Ingesters[i.id]
+		if !ok {
+			// consul must have restarted
+			log.Infof("Found empty ring, inserting tokens!")
+			ringDesc.AddIngester(i.id, i.addr, i.tokens, i.state)
+		} else {
+			ingesterDesc.Timestamp = time.Now().Unix()
+			ingesterDesc.State = i.state
+			ingesterDesc.Addr = i.addr
+			ringDesc.Ingesters[i.id] = ingesterDesc
+		}
+
+		return ringDesc, true, nil
+	})
+}
+
+// changeState updates consul with state transistions for us.  NB this must be
+// called from loop()!  Use ChangeState for calls from outside of loop().
+func (i *Ingester) changeState(state ring.IngesterState) error {
+	// Only the following state transitions can be triggered externally
+	if !((i.state == ring.PENDING && state == ring.JOINING) || // triggered by ClaimStart
+		(i.state == ring.PENDING && state == ring.ACTIVE) || // triggered by autoJoin
+		(i.state == ring.JOINING && state == ring.ACTIVE) || // triggered by ClaimFinish
+		(i.state == ring.ACTIVE && state == ring.LEAVING)) { // triggered by shutdown
+		return fmt.Errorf("Changing ingester state from %v -> %v is disallowed", i.state, state)
+	}
```

Comment:
I'd be a lot happier if this were defined as a top-level data structure `map[state][]state{}`, but it's Go so whatever.